File: src/console/command/impl/template/TemplateCommand.php
<?php

namespace pocketcloud\cloud\console\command\impl\template;

use pocketcloud\cloud\console\command\Command;
use pocketcloud\cloud\console\command\ITabComplete;
use pocketcloud\cloud\console\command\parameter\def\StringEnumParameter;
use pocketcloud\cloud\console\command\parameter\def\StringParameter;
use pocketcloud\cloud\console\command\parameter\def\TemplateParameter;
use pocketcloud\cloud\console\command\parameter\def\TemplateTypeParameter;
use pocketcloud\cloud\console\command\sender\ICommandSender;
use pocketcloud\cloud\console\command\SubCommand;
use pocketcloud\cloud\setup\def\TemplateCreationSetup;
use pocketcloud\cloud\template\Template;
use pocketcloud\cloud\template\TemplateHelper;
use pocketcloud\cloud\template\TemplateManager;
use pocketcloud\cloud\template\TemplateSettings;
use pocketcloud\cloud\template\TemplateType;
use pocketcloud\cloud\util\FormatUtils;
use pocketcloud\cloud\util\Utils;

final class TemplateCommand extends Command implements ITabComplete {

    public function __construct() {
        parent::__construct("template", "Manage your templates", ["temp"]);

        $this->registerSubCommand(SubCommand::fromClosure("create", $this->handleCreateSub(...), ["create"])
            ->addParameter(new StringParameter("name", true))
            ->addParameter(new StringParameter("templateType", true)));

        $this->registerSubCommand(SubCommand::fromClosure("edit", $this->handleEditSub(...), ["edit"])
            ->addParameter(new TemplateParameter("template", false))
            ->addParameter(new StringEnumParameter("key",  TemplateHelper::EDITABLE_KEYS, false, false))
            ->addParameter(new StringParameter("value", false)));

        $this->registerSubCommand(SubCommand::fromClosure("remove", $this->handleRemoveSub(...), ["remove"])
            ->addParameter(new TemplateParameter("template", false)));

        $this->registerSubCommand(SubCommand::fromClosure("list", $this->handleListSub(...))
            ->addParameter(new TemplateTypeParameter("type", true)));

        $this->registerSubCommand(SubCommand::fromClosure("info", $this->handleInfoSub(...))
            ->addParameter(new TemplateParameter("template", false)));
    }

    public function run(ICommandSender $sender, string $label, array $args, ?SubCommand $subCommand = null): bool {
        return true;
    }

    private function handleCreateSub(ICommandSender $sender, string $label, array $args): bool {
        if (count($args) == 0) {
            TemplateCreationSetup::new()->startSetup();
            return true;
        }

        $name = $args["name"] ?? null;
        $templateType = $args["templateType"] ?? null;
        if ($name === null || $templateType === null) return false;

        if (TemplateManager::getInstance()->check($name)) {
            $sender->error("Template already exists!");
        } else {
            TemplateManager::getInstance()->create(Template::create($name, TemplateSettings::default(), $templateType));
        }

        return true;
    }

    private function handleEditSub(ICommandSender $sender, string $label, array $args): bool {
        $template = $args["template"];
        $key = TemplateHelper::convert($args["key"]);
        $value = $args["value"];

        if (TemplateHelper::checkValue($value, $key, $expected, $realValue)) {
            TemplateManager::getInstance()->edit(
                $template,
                ($key == "lobby" ? $realValue : null),
                ($key == "maintenance" ? $realValue : null),
                ($key == "static" ? $realValue : null),
                ($key == "alwaysCopyToStaticServers" ? $realValue : null),
                ($key == "maxPlayerCount" ? $realValue : null),
                ($key == "minServerCount" ? $realValue : null),
                ($key == "maxServerCount" ? $realValue : null),
                ($key == "startNewPercentage" ? $realValue : null),
                ($key == "autoStart" ? $realValue : null),
            );
        } else $sender->error("Wrong value provided, expected an §b{} §rwhen editing §b{}§r.", $expected, $key);

        return true;
    }

    private function handleRemoveSub(ICommandSender $sender, string $label, array $args): bool {
        $template = $args["template"];
        $this->waitForConfirmation($sender, "§bAre you sure that you want to §cremove §bthe template §c" . $template->getName() . "§b?", ["yes", "y", "true", "t"])
            ->then(function (bool $confirmed) use ($sender, $template) {
                if ($confirmed) TemplateManager::getInstance()->remove($template);
            });
        return true;
    }

    public function handleListSub(ICommandSender $sender, string $label, array $args): bool {
        $type = $args["type"] ?? TemplateType::getAll();
        if (empty(TemplateManager::getInstance()->getAll(...$type))) $sender->info("§cNo templates found.");
        foreach (TemplateManager::getInstance()->getAll(...$type) as $template) {
            $sender->info(FormatUtils::implodeWithKeys(
                Utils::removeKeys($template->writeDetailed(), "statis", "alwaysCopyToStaticServers", "maxPlayerCount", "minServerCount", "maxServerCount", "startNewPercentage"),
                " §8| §r",
                "§8: §b",
                fn(string $key) => ucfirst($key),
                function (string $key, mixed $value) use($template): mixed {
                    if (in_array($key, ["lobby", "maintenance", "autoStart"])) {
                        return $value === true ? "§aYes" : "§cNo";
                    } else if ($key == "name") {
                        return $value . ($template->getParentServerGroup() !== null ? " §8(§b" . $template->getParentServerGroup()->getName() . "§8)" : "");
                    } else if ($key == "templateType") {
                        return strtoupper($value);
                    }

                    return $value;
                }
            ));
        }

        return true;
    }

    public function handleInfoSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var Template $template */
        $template = $args["template"];
        $sender->info("Information about §b{}§8:", $template->getName());
        foreach ($template->writeDetailed() as $key => $value) {
            $displayedKey = ucfirst($key);

            if (in_array($key, ["lobby", "maintenance", "static", "alwaysCopyToStaticServers", "autoStart"])) {
                $value = $value === true ? "§aYes" : "§cNo";
            } else if ($key == "name") {
                $value = $value . ($template->getParentServerGroup() !== null ? " §8(§b" . $template->getParentServerGroup()->getName() . "§8)" : "");
            } else if ($key == "startNewPercentage") {
                $value = $value . "%";
            } else if ($key == "templateType") {
                $value = strtoupper($value);
            }

            $sender->info($displayedKey . "§8: §b" . $value);
        }

        return true;
    }

    public function onTabComplete(array $args): array {
        if (count($args) == 4) {
            if ($args[0] == "edit") {
                $selectedKey = TemplateHelper::convert($args[2]);
                if (TemplateHelper::checkKey($selectedKey) && in_array($selectedKey, ["lobby", "maintenance", "autostart", "static", "alwayscopytostaticservers"])) {
                    return ["true", "false"];
                }
            }
        }
        return [];
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/impl/template/TemplateCommand.java
package de.pocketcloud.cloud.console.command.impl.template;

import de.pocketcloud.api.component.template.ITemplate;
import de.pocketcloud.api.search.TemplateSearchQuery;
import de.pocketcloud.api.template.TemplateType;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.command.Command;
import de.pocketcloud.cloud.console.command.TabComplete;
import de.pocketcloud.cloud.console.command.ctx.CommandContext;
import de.pocketcloud.cloud.console.command.desc.CommandDescription;
import de.pocketcloud.cloud.console.command.flag.CommandFlag;
import de.pocketcloud.cloud.console.command.parameter.def.*;
import de.pocketcloud.cloud.console.command.sender.CommandSender;
import de.pocketcloud.cloud.console.command.sub.SubCommand;
import de.pocketcloud.cloud.setup.def.TemplateCreationSetup;
import de.pocketcloud.cloud.setup.def.TemplateEditSetup;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.cloud.template.util.TemplateHelper;
import de.pocketcloud.shared.component.software.ServerSoftware;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;

@CommandDescription(name = "template", description = "Manage the templates", aliases = {"temp"})
public final class TemplateCommand extends Command implements TabComplete {

    @Override
    public void prepare() {
        registerSubCommand(SubCommand.lambda("create", this::handleCreateSub, true));

        registerSubCommand(SubCommand.lambda("remove", this::handleRemoveSub, true, subCommand -> {
            subCommand.addParameter(new TemplateParameter("template", false))
                    .addFlag(CommandFlag.shortFlag("y"));
        }));

        registerSubCommand(SubCommand.lambda("edit", this::handleEditSub, true, subCommand -> {
            subCommand.addParameter(new TemplateParameter("template", false));
        }));

        registerSubCommand(SubCommand.lambda("info", this::handleInfoSub, true, subCommand -> {
            subCommand.addParameter(new TemplateParameter("template", false));
        }));

        registerSubCommand(SubCommand.lambda("list", this::handleListSub, true, subCommand -> {
            subCommand.addParameter(new MultipleTypesParameter("filter", true, new TemplateTypeParameter("template_type", true),
                    new ServerGroupParameter("server_group", true),
                    new SoftwareParameter("software", true)));
        }));
    }

    @Override
    public boolean run(CommandSender sender, CommandContext ctx) {
        return true;
    }

    public boolean handleCreateSub(CommandSender sender, CommandContext ctx) {
        new TemplateCreationSetup().startSetup();
        return true;
    }

    public boolean handleRemoveSub(CommandSender sender, CommandContext ctx) {
        Template template = ctx.arg("template", Template.class);
        boolean sure = ctx.hasFlag("y");
        if (sure) {
            PocketCloud.instance().templates().remove(template);
        } else {
            awaitConfirmation(sender, "Are you sure you want to §cremove §rthe template §b" + template.name() + "§r?")
                    .thenSuccess(r -> {
                        if (r) {
                            PocketCloud.instance().templates().delete(template);
                        }
                    });
        }
        return true;
    }

    public boolean handleEditSub(CommandSender sender, CommandContext ctx) {
        Template template = ctx.arg("template", Template.class);
        new TemplateEditSetup(template).startSetup();
        return true;
    }

    public boolean handleInfoSub(CommandSender sender, CommandContext ctx) {
        Template template = ctx.arg("template", Template.class);
        sender.info("Name§8: §b{}", template.name());
        sender.info("Lobby§8: §b{}", template.settings().lobby() ? "§aYes" : "§cNo");
        sender.info("Maintenance§8: §b{}", template.settings().maintenance() ? "§aYes" : "§cNo");
        sender.info("Static§8: §b{}", template.settings().staticServers() ? "§aYes" : "§cNo");
        sender.info("AlwaysCopyToStaticServers§8: §b{}", template.settings().alwaysCopyToStaticServers() ? "§aYes" : "§cNo");
        sender.info("SaveOnShutdown§8: §b{}", template.settings().saveOnShutdown() ? "§aYes" : "§cNo");
        sender.info("MaxPlayerCount§8: §b{} player(s)", template.settings().maxPlayerCount());
        sender.info("MinServerCount§8: §b{} server(s)", template.settings().minServerCount());
        sender.info("MaxServerCount§8: §b{} server(s)", template.settings().maxServerCount());
        sender.info("StartNewPercentage§8: §b{}%", template.settings().startNewPercentage() * 100);
        sender.info("AutoStart§8: §b{}", template.settings().autoStart() ? "§aYes" : "§cNo");
        sender.info("Software§8: §b{} §8(§b{}§8)", template.serverSoftware().name(), template.templateType().name());
        sender.info("§b{} player(s) §racross §b{} server(s)§r.", template.playerCount(), template.serverCount());
        return true;
    }

    public boolean handleListSub(CommandSender sender, CommandContext ctx) {
        TemplateSearchQuery query = TemplateSearchQuery.create();
        Object filter = ctx.arg("filter");
        if (filter instanceof TemplateType type) {
            query.ofTemplateType(type);
        } else if (filter instanceof ServerGroup group) {
            query.ofServerGroup(group);
        } else if (filter instanceof ServerSoftware software) {
            query.runningSoftware(software.name());
        }

        Collection<ITemplate> templates = PocketCloud.instance().templates().query(query).stream().sorted(Comparator.comparingInt(ITemplate::playerCount)).toList();
        sender.info("Templates §8(§b{}§8/§b{}§8)§r:", templates.size(), PocketCloud.instance().templates().getAll().size());
        if (templates.isEmpty()) sender.info("§cNo templates found.");
        for (ITemplate template : templates) {
            sender.info("Name§8: §b{} §8| §rPlayers§8: §b{}§8/§c{} §8| §rServers§8: §b{}§8/§c{} §8| §rLobby§8: §b{} §8| §rMaintenance§8: §b{} §8| §rSoftware§8: §b{} §8(§b{}§8)",
                    template.name(), template.playerCount(), template.settings().maxPlayerCount(), template.serverCount(), template.settings().maxServerCount(),
                    template.settings().lobby() ? "§aYes" : "§cNo", template.settings().maintenance() ? "§aYes" : "§cNo",
                    template.serverSoftware().name(), template.templateType().name());
        }

        return true;
    }

    @Override
    public Collection<String> onTabComplete(List<String> args) {
        if (args.size() == 4) {
            // template (/) edit (0) Lobby-1 (1) autoStart (2) false (3)
            if (args.getFirst().equals("edit")) {
                String key = TemplateHelper.convert(args.get(2));
                if (TemplateHelper.checkKey(key) && TemplateHelper.getKeyType(key) == Boolean.class) {
                    return List.of("true", "false");
                }
            }
        }
        return List.of();
    }
}